src/ctia/schemas/utils.clj

```diff
@@ -18,55 +19,169 @@
+                       (pr-str service-kw))
+               (assert (simple-keyword? (s/explicit-schema-key service-kw))
+                       (s/explicit-schema-key service-kw))
+               (assert (set? fn-kws))
+               (let [gval (get graph-value (s/explicit-schema-key service-kw))]
+                 (if-not (map? gval)
+                   (when-not (s/optional-key? service-kw)
+                     (throw (ex-info (str "Missing service: " service-kw) {})))
+                   (let [service-fns (into {}
+                                           (map (fn [fn-kw]
+                                                  (assert (simple-keyword? (s/explicit-schema-key fn-kw))
+                                                          (s/explicit-schema-key fn-kw))
+                                                  (if-some [svc-fn (get gval (s/explicit-schema-key fn-kw))]
+                                                    {(s/explicit-schema-key fn-kw) svc-fn}
+                                                    (when-not (s/optional-key? fn-kw)
+                                                      (throw (ex-info (str "Missing " service-kw " service function: "
+                                                                           (s/explicit-schema-key fn-kw))
+                                                                      {}))))))
+                                           fn-kws)]
+                     {(s/explicit-schema-key service-kw) service-fns})))))
+        selectors))
+
+(s/defn service-subgraph-from-schema :- (s/pred map?)
+  "Given a schema describing a Trapperkeeper graph,
+  returns just the elements in graph mentioned
+  in the schema as 'explicit keys' using service-subgraph.
+  Throws an exception if any levels are missing.
+  
+  (service-subgraph-from-schema
+    {:ConfigService {:get-in-config (fn [...] ...)
+                     :get-config (fn [...] ...)}
+     :FooService {:f1 (fn [...] ...)
+                  :f2 (fn [...] ...)}
+     :BarService {:b1 (fn [...] ...)}}
+    {:ConfigService {:get-in-config (s/=> ...)}
+     (s/optional-key :FooService) {:f1 (s/=> ...)
+                                   (s/optional-key :f2) (s/=> ...)}})
+  ;=> {:ConfigService {:get-in-config (fn [...] ...)}
+  ;    :FooService {:f1 (fn [...] ...)
+  ;                 :f2 (fn [...] ...)}}
+  "
+  [graph-value :- (s/pred map?)
+   schema :- (s/protocol s/Schema)]
+  {:pre [(map? schema)]}
+  (service-subgraph
+    graph-value
+    ;; this could be factored out a la schema.coerce/coercer for
+    ;; better performance.
+    (into {}
+          (map (fn [[service-kw service-fns]]
+                 {:pre [(map? service-fns)]}
+                 (when (s/specific-key? service-kw)
+                   {service-kw
+                    (into #{}
+                          (filter s/specific-key?)
+                          (keys service-fns))})))
+          schema)))
+
+(s/defn service-subschema :- (s/protocol s/Schema)
```

Comment:
hmmm I think I would use `service-schema-select` to capture the idea that this is not "check if service-schema-1 is a sub-schema of service-schema-2".

Reply:
Thanks, added a `select-*` prefix to all the functions.